### DISCUSSION TITLE
curl_easy_perform(curl)   crashed

### DISCUSSION THREAD

**Discussion by u32744** [curl/curl#15569] [Q&A] [answered]
crashed at this code:
` result = events ? easy_events(multi) : easy_transfer(multi);    `

source code:
~~~c
size_t callback_get_head(void* ptr, size_t size, size_t nmemb, void* userp) {
    char* str = (char*)userp;
    if (str == nullptr) {
        return 0; // 如果 userp 是 NULL，则不进行任何操作
    }

    size_t len = strlen(str);
    size_t newlen = len + size * nmemb;

    // 确保有足够的空间来存储新的数据
    if (newlen > len) {
        str = (char*)realloc(str, newlen + 1); // +1 for null terminator
        if (str == nullptr) {
            return 0; // realloc 失败
        }
        userp = str; // 更新 userp 到新分配的内存
    }

    memcpy(str + len, ptr, size * nmemb);
    str[newlen] = '\0'; // 添加 null 终止符

    return size * nmemb;
}

static char connect_cloud(char* pc_ret, const char* host_addr, const int portno,
    const char* pc_method, struct curl_slist* http_headers) {
    assert((pc_ret != NULL) && (host_addr != NULL) && (pc_method != NULL));

    CURL* curl = curl_easy_init();
    if (!curl) {
        fprintf(stderr, "--- curl init Error!\n");
        return 0;
    }

    curl_easy_setopt(curl, CURLOPT_URL, host_addr);
    curl_easy_setopt(curl, CURLOPT_HTTP_VERSION, CURL_HTTP_VERSION_1_1);
    curl_easy_setopt(curl, CURLOPT_TIMEOUT, 63);     // 设置超时，单位为秒
    curl_easy_setopt(curl, CURLOPT_NOSIGNAL, 1L);    // 屏蔽其它信号，使得 libcurl 不会忽略 SIGPIPE 信号
    curl_easy_setopt(curl, CURLOPT_HEADER, 1L);      // 下载数据包括 HTTP 头部，默认值为 0

#if (CURL_DBG == 1)
    curl_easy_setopt(curl, CURLOPT_VERBOSE, 1L);     // 调试用：显示交互明细
#endif

    curl_easy_setopt(curl, CURLOPT_CUSTOMREQUEST, pc_method);
    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, callback_get_head); // 下载数据的回调函数
    curl_easy_setopt(curl, CURLOPT_WRITEDATA, pc_ret);                // 下载数据的指针
    curl_easy_setopt(curl, CURLOPT_HTTPHEADER, http_headers);         // 多行的 HTTP 头部
    curl_easy_setopt(curl, CURLOPT_IPRESOLVE, CURL_IPRESOLVE_V4);      // 强制使用 IPv4

    CURLcode res = curl_easy_perform(curl);
    if (res != CURLE_OK) {
        fprintf(stderr, "curl_easy_perform() failed: %s\n", curl_easy_strerror(res));
        curl_easy_cleanup(curl);
        return 0;
    }

    curl_easy_cleanup(curl);

**Comment by dfandrich** [comment, marked as answer]
You seem to have a fundamental misunderstanding about the use of C pointers. 

> `size_t callback_get_head(void* ptr, size_t size, size_t nmemb, void* userp) {
> char* str = (char*)userp;

This casts the incoming pointer as a pointer to char, which is fine.

> if (str == nullptr) {
> return 0; // 如果 userp 是 NULL，则不进行任何操作
> }
> 
> size_t len = strlen(str);
> size_t newlen = len + size * nmemb;
> 
> // 确保有足够的空间来存储新的数据
> if (newlen > len) {
>     str = (char*)realloc(str, newlen + 1); // +1 for null terminator

This is suspicious but not necessarily guaranteed to be wrong in isolation.
realloc will free the str pointer and allocate a new block if it's not able to
change the size in place. So, str at this point is likely different from userp.

>     if (str == nullptr) {
>         return 0; // realloc 失败
>     }
>     userp = str; // 更新 userp 到新分配的内存

At this point, it's finally clear that this is incorrect. userp is a local
variable. Assigning to it has no effect on the caller. The new pointer
allocated in realloc is then lost at the end of this function and the original
one has been freed. The next call to the function will be passing in a freed
pointer which is garbage to the realloc call and a crash is likely.

> }
> 
> memcpy(str + len, ptr, size * nmemb);
> str[newlen] = '\0'; // 添加 null 终止符
> 
> return size * nmemb;
> 
> }
